improving interval arithmetic for some special cases

michaelbynum opened the pull request:
## Summary/Motivation:
The ultimate goal of this PR is for 
```
import pyomo.environ as pe
from pyomo.contrib.fbbt.fbbt import compute_bounds_on_expr

m = pe.ConcreteModel()
m.x = pe.Var()
m.y = pe.Var()
e = (m.x**2 + m.y**2)**(-0.5)
compute_bounds_on_expr(e)
```

to produce

```
(0, None)
```

instead of 

```
(None, None)
```

Both of these are valid, but the former is much more useful. In the end, this took a rework of the interval `div` and `inv` functions. 



### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

jsiirola commented on pyomo/contrib/fbbt/interval.py:
```diff
@@ -39,35 +39,59 @@ def mul(xl, xu, yl, yu):
         If x is always positive, things are simple. We only need to worry about the sign of y.
         """
-        if yl < 0 and yu > 0:
+        if yl < 0 < yu:
```
This case's else branch is unreachable.  Consider changing to:
```python
else: # yu <= 0:
    lb = min(xu**yl, xu**yu)
    ub = max(xl**yl, xl**yu)
```

michaelbynum replied:
Good point.

michaelbynum replied:
fixed

jsiirola commented on pyomo/contrib/fbbt/interval.py:
```diff
@@ -39,35 +39,59 @@ def mul(xl, xu, yl, yu):
+    """
+    if xu <= 0 <= xl:
+        if abs(xl - xu) <= feasibility_tol:
+            raise IntervalException(f'Division by zero in inv; xl: {xl}; xu: {xu}')
```
Is this correct?  This logic seems a bit strange: You allow for division by zero for the intervals `[-eps, eps]`, `[eps, n]`, and `[-n, eps]`, but not `[eps, -eps]`?

michaelbynum replied:
I believe this behavior is reasonable, but it might be safer to return `(-inf, inf)` for this case.

michaelbynum replied:
This one is tricky. I will come back to it. We definitely need to be careful here.

jsiirola replied:
I see.  That makes sense.  Would this code then be more clear if it was written as 
```python
if xu - xl <= -feasibility_tol:
    raise InfeasibleConstraintException
elif xu - xl <= feasibility_tol and min(xl, xu) <= 0 <= max(xl, xu):
    raise IntervalException(f'Division by zero in inv')
```

michaelbynum replied:
Yes. I like that. I'll make that change shortly.

michaelbynum replied:
Actually, I think it has to be 

```
if xu - xl <= -feasibility_tol:
    raise InfeasibleConstraintException
elif xu <= 0 <= xl:
    raise IntervalException(f'Division by zero in inv')
```

because I want `inv(-1e-7, 1e-7, feasibility_tol=1e-6)` to return `(-inf, inf)` and not raise an exception.

michaelbynum replied:
I updated this part of the code.

jsiirola commented on pyomo/contrib/fbbt/interval.py:
```diff
@@ -39,35 +39,59 @@ def mul(xl, xu, yl, yu):
+        if abs(xl - xu) <= feasibility_tol:
+            raise IntervalException(f'Division by zero in inv; xl: {xl}; xu: {xu}')
+        else:
+            raise InfeasibleConstraintException(f'lower bound is greater than upper bound in inv; xl: {xl}; xu: {xu}')
```
Why do you only raise this exception if `xu < xl` *and* spans 0?  For example, the interval `[3, 2]` would not trigger this exception?

michaelbynum replied:
The interval `[3, 2]` would raise an exception, but only when performing FBBT. That is where most of the infeasible constraint detection happens.

jsiirola commented on pyomo/contrib/fbbt/interval.py:
```diff
@@ -39,35 +39,59 @@ def mul(xl, xu, yl, yu):
         lb = 1.0 / xu
+    else:
+        # everything else
+        lb = -inf
+        ub = inf
```
Why does `[-eps, 2]` return `[-inf, inf]`, but `[eps, 2]` returns `[0.5, inf]`?

michaelbynum replied:
This is just me being careful. If the lower bound is negative at all, I return `-inf` for the lower bound (1/-1e-12 is -1e12). If `xl >= -eps`, it could reasonably be assumed that the result should positive. However, it is safer to return `-inf` for the resulting lower bound.

jsiirola commented on pyomo/contrib/fbbt/interval.py:
```diff
@@ -82,9 +106,13 @@ def power(xl, xu, yl, yu):
+            lb1, ub1 = power(xl, xu, 0, yu, feasibility_tol)
+            lb2, ub2 = power(xl, xu, yl, 0, feasibility_tol)
+            lb = min(lb1, lb2)
+            ub = max(ub1, ub2)
```
Why is the next clause not updated to be relaxed by `feasibility_tol`?

michaelbynum replied:
Can you be more specific? Which part do you think needs relaxed?

michaelbynum replied:
Nevermind. I did not see your comment below.

jsiirola commented on pyomo/contrib/fbbt/interval.py:
```diff
@@ -140,7 +168,7 @@ def power(xl, xu, yl, yu):
             raise InfeasibleConstraintException(msg)
         xl = 0
-        lb, ub = power(xl, xu, yl, yu)
+        lb, ub = power(xl, xu, yl, yu, feasibility_tol)
```
Why is this clause (including `yl == yu` above) not relaxed by `feasibility_tol`?

michaelbynum replied:
The purpose of this clause is to catch constant exponents only (e.g., `m.x**1.5`). We do not want to catch things like `m.x**m.y` where `1 <= m.y <= 1+1e-4`. In that case, the `m.y` is not guaranteed to be fractional, so we cannot assert that m.x has to be positive. Relaxing `yl == yu` with `feasibility_tol` is actually not a relaxation.

jsiirola reviewed (commented):
There are several nuances here that I am not following.  I understand why you are propagating the `feasibility_tol` into these methods, but I don't think you are covering all the cases, and I thing there are more of these methods that should also be "relaxed" by the `feasibility_tol`.

michaelbynum commented:
I also updated the FBBT tests with long runtimes. I was able to utilize `np.split` to remove some of the `for` loops. On my machine this brought the total runtime of all FBBT tests from 12 seconds to 2 seconds.

michaelbynum commented on pyomo/contrib/fbbt/interval.py:
```diff
@@ -39,52 +39,77 @@ def mul(xl, xu, yl, yu):
 def div(xl, xu, yl, yu, feasibility_tol):
-    if yl <= feasibility_tol and yu >= -feasibility_tol:
-        # if the denominator (y) can include 0, then x/y is unbounded.
+    if xl <= 0 <= xu and yl <= 0 <= yu:
         lb = -inf
         ub = inf
```
No, this is still needed. This special case indicates that we don't know anything if `0/0` is possible.
